src/unix/signal.c

```diff
@@ -457,8 +498,17 @@ static void uv__signal_stop(uv_signal_t* handle) {
   /* Check if there are other active signal watchers observing this signal. If
    * not, unregister the signal handler.
    */
-  if (uv__signal_first_handle(handle->signum) == NULL)
+  first_handle = uv__signal_first_handle(handle->signum);
+  if (first_handle == NULL) {
     uv__signal_unregister_handler(handle->signum);
+  } else {
+    rem_one_shot = handle->flags & UV__SIGNAL_ONE_SHOT;
+    first_one_shot = first_handle->flags & UV__SIGNAL_ONE_SHOT;
+    if (first_one_shot && !rem_one_shot)
+      assert(uv__signal_register_handler(handle->signum, 1) == 0);
+    else if (!first_one_shot && rem_one_shot)
+      assert(uv__signal_register_handler(handle->signum, 0) == 0);
```

Comment:
You could simplify this to:
```c
if (first_one_shot ^ rem_one_shot)
  assert(uv__signal_register_handler(handle->signum, first_one_shot) == 0);
```
Although you could debate whether it's more legible.

At any rate, `uv__signal_register_handler()` has side effects so don't call it from inside the assert statement; it's going to get compiled out of optimized (`-DNDEBUG`) builds.